File: ai-service/app/domain/booking.py
from enum import StrEnum
from typing import Any, Literal

from pydantic import BaseModel, Field

from app.domain.travel import TripDetails


class BookingProductType(StrEnum):
    TRAIN = "train"
    FLIGHT = "flight"
    BUS = "bus"
    HOTEL = "hotel"


class BookingStep(StrEnum):
    SELECT_CARRIAGE = "select_carriage"
    SELECT_ROOM = "select_room"
    SELECT_FARE = "select_fare"
    SELECT_EXTRAS = "select_extras"
    SELECT_SEATS = "select_seats"
    CONFIRM_FARE = "confirm_fare"
    PASSENGERS = "passengers"
    DOCUMENTS = "documents"
    GUESTS = "guests"
    CONFIRM = "confirm"
    CHECKOUT = "checkout"


class BookingOption(BaseModel):
    id: str
    title: str
    description: str
    price_delta: int = 0
    available: bool = True


class BookingTravelerDraft(BaseModel):
    full_name: str | None = None
    birth_date: str | None = None
    document_type: Literal[
        "international_passport",
        "domestic_passport",
        "birth_certificate",
    ] | None = None
    document_number: str | None = None


class BookingCopilotDecision(BaseModel):
    assistant_message: str = Field(min_length=1, max_length=2_000)
    option_id: str | None = Field(
        default=None,
        description="Exact available option id for a single-choice step.",
    )
    option_ids: list[str] = Field(
        default_factory=list,
        description="Exact available option ids for the select_extras step.",
    )
    seat_ids: list[str] = Field(
        default_factory=list,
        description="Exact available seat ids, one per traveler, for select_seats.",
    )
    travelers: list[BookingTravelerDraft] = Field(
        default_factory=list,
        description="Passenger or guest drafts copied only from explicit user-provided facts.",
    )
    missing_fields: list[str] = Field(
        default_factory=list,
        description="Fields that are still unknown and must be provided by the user.",
    )


class BookingCopilotRequest(BaseModel):
    product_type: BookingProductType
    current_step: BookingStep
    travelers_count: int = Field(ge=1, le=20)
    current_options: list[BookingOption] = Field(default_factory=list)
    selections: dict[str, Any] = Field(default_factory=dict)
    trip: TripDetails
    conversation: list[str] = Field(default_factory=list, max_length=200)
    instruction: str = Field(default="", max_length=5_000)


class BookingCopilotResponse(BaseModel):
    assistant_message: str
    proposed_data: dict[str, Any] = Field(default_factory=dict)
    missing_fields: list[str] = Field(default_factory=list)
    can_apply: bool = False
    requires_user_confirmation: bool = True

File: frontend/src/api/booking.ts
import type { SearchOption } from "./chat";

export type BookingProductType = "train" | "flight" | "bus" | "hotel";

export type BookingStep =
  | "select_carriage"
  | "select_room"
  | "select_fare"
  | "select_extras"
  | "select_seats"
  | "confirm_fare"
  | "passengers"
  | "documents"
  | "guests"
  | "confirm"
  | "checkout";

export interface BookingStepOption {
  id: string;
  title: string;
  description: string;
  price_delta: number;
  available: boolean;
}

export interface Booking {
  id: string;
  user_id: string;
  session_id: string;
  product_type: BookingProductType;
  option: SearchOption;
  steps: BookingStep[];
  current_step: BookingStep;
  completed_steps: BookingStep[];
  selections: Record<string, unknown>;
  travelers_count: number;
  current_options: BookingStepOption[];
  checkout_url: string | null;
  inventory_source: "preview";
  provider_notice: string;
}

export interface BookingAssistance {
  assistant_message: string;
  proposed_data: Record<string, unknown>;
  missing_fields: string[];
  can_apply: boolean;
  requires_user_confirmation: boolean;
}

export interface ExtractedPassengerDocument {
  document_type: "international_passport" | "domestic_passport" | "birth_certificate" | "unknown";
  last_name: string | null;
  first_name: string | null;
  middle_name: string | null;
  last_name_latin: string | null;
  first_name_latin: string | null;
  date_of_birth: string | null;
  document_series: string | null;
  document_number: string | null;
}

export interface DocumentExtractionResult {
  document: ExtractedPassengerDocument;
  missing_fields: string[];
  manual_review_required: boolean;
}

interface ApiErrorPayload {
  detail?: string;
}

export async function createBooking(request: {
  user_id: string;
  session_id: string;
  option: SearchOption;
}): Promise<Booking> {
  return bookingRequest("/api/v1/bookings", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(request),
  });
}

export async function getBooking(bookingId: string, userId: string): Promise<Booking> {
  return bookingRequest(
    `/api/v1/bookings/${encodeURIComponent(bookingId)}?user_id=${encodeURIComponent(userId)}`,
  );
}

export async function submitBookingStep(
  bookingId: string,
  request: { user_id: string; step: BookingStep; data: Record<string, unknown> },
): Promise<Booking> {
  return bookingRequest(`/api/v1/bookings/${encodeURIComponent(bookingId)}/steps`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(request),
  });
}

export async function assistBooking(
  bookingId: string,
  request: { user_id: string; instruction: string },
): Promise<BookingAssistance> {
  return apiRequest(`/api/v1/bookings/${encodeURIComponent(bookingId)}/assist`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(request),
  });
}

export async function extractPassengerDocument(
  userId: string,
  sessionId: string,
  file: File,
): Promise<DocumentExtractionResult> {
  const formData = new FormData();
  formData.append("document", file);
  return apiRequest(
    `/api/v1/agent/users/${encodeURIComponent(userId)}/sessions/${encodeURIComponent(sessionId)}/documents/extract`,
    { method: "POST", body: formData },
  );
}

async function apiRequest<T>(url: string, init?: RequestInit): Promise<T> {
  const response = await fetch(url, { credentials: "include", ...init });
  if (!response.ok) {
    const payload = (await response.json().catch(() => ({}))) as ApiErrorPayload;
    throw new Error(payload.detail ?? "Не удалось продолжить оформление.");
  }
  return response.json() as Promise<T>;
}

async function bookingRequest(url: string, init?: RequestInit): Promise<Booking> {
  return apiRequest<Booking>(url, init);
}
